Objects/dictobject.c

```diff
@@ -6718,6 +6771,66 @@ _PyObject_StoreInstanceAttribute(PyObject *obj, PyDictValues *values,
     return 0;
 }
 
+static inline int
+store_instance_attr_dict(PyObject *obj, PyDictObject *dict, PyObject *name, PyObject *value)
+{
+    PyDictValues *values = _PyObject_InlineValues(obj);
+    int res;
+    Py_BEGIN_CRITICAL_SECTION(dict);
+    if (dict->ma_values == values) {
+        res = store_instance_attr_lock_held(obj, values, name, value);
+    }
+    else {
+        if (value == NULL) {
+            res = PyDict_DelItem((PyObject *)dict, name);
+        } else {
+            res = PyDict_SetItem((PyObject *)dict, name, value);
+        }
+    }
+    Py_END_CRITICAL_SECTION();
+    return res;
+}
+
+int
+_PyObject_StoreInstanceAttribute(PyObject *obj, PyObject *name, PyObject *value)
+{
+    PyDictValues *values = _PyObject_InlineValues(obj);
+    if (!FT_ATOMIC_LOAD_UINT8_RELAXED(values->valid)) {
+        return store_instance_attr_dict(obj, _PyObject_GetManagedDict(obj), name, value);
+    }
```

Comment:
`_PyObject_GetManagedDict(obj)` can be `NULL` here:

The following crashes (we should add it or something similar as a unit test):

```python
class Foo:
    pass

f = Foo()
del f.__dict__
f.a = 3
```